### DISCUSSION TITLE
Error: Couldn't find baseline in commit (using vcpkg registries)

### DISCUSSION THREAD

**Discussion by drathin4** [microsoft/vcpkg#27308] [Q&A]
I'm trying to add a custom library (closed-source) to VCPkg using the manifests & registries feature on Windows platform. I have _vcpkg-configuration.json_ file in the root directory of the project whose contents are:

{ "registries": [ { "kind": "git", "repository": "git@ssh.dev.azure.com:v3/neilsoft-innovation/scan2bim/finitor", "baseline": "1f5fdd4155ff86e9629399c9713faa81ba9e7093", "packages": [ "fleck" ] } ] }

and a _vcpkg.json_ file whose contents are:

{ "name": "fleck_package", "description": "A point-cloud engine tailored for Scan2BIM project", "homepage": "https://dev.azure.com/neilsoft-innovation/scan2bim/_git/finitor?version=GBfleck_python_binding&path=/fleck", "version": "1.0", "dependencies": [ "vectorclass", "eigen3", "pcl", "qt5", "fleck" { "name": "vcpkg-cmake", "host": true }, { "name": "vcpkg-cmake-config", "host": true } ], "$platform": "(windows & x64)" }

The _baseline_ field in the _vcpkg-configuration.json_ file is the git-sha commit id of _baseline.json_ file.

The _vcpkg-registry_ directory structure at the azure location is:

vcpkg-registry/ports/fleck/portfile.cmake
vcpkg-registry/ports/fleck/vcpkg.json
vcpkg-registry/versions/f-/fleck.json
vcpkg-registry/versions/baseline.json

When I enter the command `vcpkg install --feature-flags=registries,manifests` in CMD, it gives the error:

**Error: Couldn't find baseline in commit `"1f5fdd4155ff86e9629399c9713faa81ba9e7093"` from repo git@ssh.dev.azure.com:v3/neilsoft-innovation/scan2bim/finitor:
error: git failed with exit code: (128).
fatal: path 'versions/baseline.json' exists on disk, but not in '1f5fdd4155ff86e9629399c9713faa81ba9e7093'**

I used the git-sha `1f5fdd4155ff86e9629399c9713faa81ba9e7093` using the commit id of _baseline.json_ file by examining the commit history on azure. So why is vcpkg complaining about not finding this _baseline.json_ file? Has it something to do with my branch because I'm not using master/main branch ?

A few points that I'd like to mention before closing this query.
1. I found a discussion of that error on [https://github.com/microsoft/vcpkg/issues/23748](url), but that again is different (I'm using registries and not versions and that too on Windows platform).
2. I have made my branch on azure as the default branch instead of main/master branch, but despite of that I'm getting this error.
3. In one of the blog posts, I read that the _repository_ field in _vcpkg-configuration.json_ is full path to the _vcpkg-registry_ directory, but I couldn't include '_/vcpkg-registry_' in that field, else `vcpkg install --feature-flags=registries,manifests` command fails instantly.

**Comment by strega-nil-ms** [comment]
It looks like the issue is that `finitor` doesn't have `versions` at the base of the repository; the way that vcpkg registries are designed, the registry must be its own git repository.

**Reply by drathin4** [reply]
Earlier, I had created a directory `vcpkg-registry` in the repository `finitor` under my branch. But as per your suggestion, I created a new repository itself called `vcpkg-registry` with the same directory structure as mentioned above. Now, the `versions` & `ports` folders are directly in the root folder of the repository.

The `vcpkg-configuration.json` file is:

{
  "registries": [
    {
      "kind": "git",
      "repository": "git@ssh.dev.azure.com:v3/neilsoft-innovation/scan2bim/vcpkg-registry",
      "baseline": "b3726a7c96b724b723db5dd2915069f04475a084",
      "packages": [
        "fleck"
      ]
    }
  ]
}

The `vcpkg.json` file is unchanged:

{
  "name": "fleck-package",
  "description": "A point-cloud engine tailored for Scan2BIM project",
  "homepage": "https://dev.azure.com/neilsoft-innovation/scan2bim/_git/finitor?version=GBfleck_python_binding&path=/fleck",
  "version": "1.0",
  "dependencies": [
    "vectorclass",
    "eigen3",
    "pcl",
    "qt5",
    "fleck",
    {
      "name": "vcpkg-cmake",
      "host": true
    },
    {
      "name": "vcpkg-cmake-config",
      "host": true
    }
  ],
  "$platform": "(windows & x64)"
}

The library `fleck` that I want to integrate in VCPkg is under `fleck_python_binding` branch of the repository `finitor`. Also, the `baseline` field of `vcpkg-configuration.json` is the git-sha of the file `baseline.json` i.e. `b3726a7c96b724b723db5dd2915069f04475a084`

The command `vcpkg install --feature-flags=registries,manifests`  gives the following error:

**error: while loading b3726a7c96b724b723db5dd2915069f04475a084:
Failed to find either a CONTROL file or vcpkg.json file.Error: Failed to load port fleck from C:\Users\SEZ5505\AppData\Local\vcpkg\registries\git-trees\b3726a7c96b724b723db5dd2915069f04475a084
note: updating vcpkg by rerunning bootstrap-vcpkg may resolve this failure.**

All the files are present in `vcpkg-registry` repository including `vcpkg.json`, but still VCPkg is complaining. What could be the issue ?

**Reply by strega-nil-ms** [reply]
What does `versions/f-/fleck.json` look like?

**Reply by drathin4** [reply]
`versions/f-/fleck.json:`

{
	"versions": [
		{
			"version": "1.0",
            "port-version": 0,
			"git-tree": "b3726a7c96b724b723db5dd2915069f04475a084"
		}
	]
}

**Reply by drathin4** [reply]
I suspect, I've made a mistake in providing the `git-tree` identifier. I've used the same commit-id of `versions/baseline.json`.

In the example, I've seen the following:
`
git add ports/beicode -> 
git commit -m "[beicode] new port" -> 
git rev-parse HEAD:ports/beicode -> 
tree identifier: 7fb5482270b093d40ab8ac31db89da4f880f01ba
`

As per my understanding that is the identifier of the `directory` and not the `file`. Since I'm using Azure and creating the files directly on cloud (rather than through command line), I don't know how to get the directory identifier.

Kindly let me know if I've entered the wrong `git-tree` identifier in `fleck.json.`

**Reply by strega-nil-ms** [reply]
I have no idea how to do this on the Azure UI. I would recommend `clone`-ing the repository and doing the work there.

**Reply by BillyONeal** [reply]
https://gist.github.com/BillyONeal/04a77c971d87a5b0e3727a42dd9e5dbe

I agree with Nicole, I don't know practical ways of doing this without using git itself

**Reply by JavierMatosD** [reply]
The issue is that the "git-tree" field should be pointing to the git tree of the port directory (i.e. ports/fleck).
You can get the sha by running the git command "ls-tree" on the port's directory. 
For example, x-add-version does this: 

``` 
"C:\Program Files\Git\cmd\git.exe" -c core.autocrlf=false -C "C:\dev\vcpkg\ports" ls-tree -d HEAD --
```

>I'm trying to add a custom library (closed-source) to VCPkg using the manifests & registries feature on Windows platform. 

You should make your registry using x-add-version. :)

**Reply by drathin4** [reply]
Thanks for your replies @strega-nil-ms & @JavierMatosD. I was able to get past the initial `VCPkg` error using plain `Git` commands, but I'm getting a `CMake` error. I have to make some changes in `CMakeLists.txt` file of the `fleck` library. If I face some `VCPkg` error again, I'll reply in this thread. Thanks again.
